### ISSUE TITLE
Implicit Type Not Passing Through Generic With String Map

### ISSUE THREAD

**Issue by sam-s4s** [microsoft/TypeScript#33000]
**TypeScript Version:**  3.5.3

**Search Terms:** implicit type passing string map generic

**Code**

```ts
// A *self-contained* demonstration of the problem follows...
// Test this by running `tsc` on the command-line, rather than through another build tool such as Gulp, Webpack, etc.
export interface ICrxModelEvent<T> {
    'change': ValueCallback<T>;
    'changeInternal': ValueCallback<T>;
    'changeExternal': ValueCallback<T>;
}

type ValueCallback<T> = (value: T) => void;

export abstract class CrxModel<T> {
    addEvent<K extends keyof ICrxModelEvent<T>>(type: K, callback: ICrxModelEvent<T>[K]) {

    }
}

export class CrxNumber extends CrxModel<number> { }
```
and then to test it...
```ts
let num = new CrxNumber();
num.addEvent('change', v => {});
```

**Expected behavior:**
param `v` is implicitly typed as `number`.

**Actual behavior:**
param `v` has implicit type of `any`, (which, as a side-note, in my case give a no implicit any error)

(and when I mouseover the `addEvent` call, it has correct implicit types listed for both `'change'` and `number`)

Note that if I specify the type explicitly, it works, and `v` is correctly types as `number`:
```ts
num.addEvent<'change'>('change', v => {});
```

**Additional Info:**

The idea is basically an extension of `lib.dom.d.ts`:
```ts
addEventListener<K extends keyof HTMLElementEventMap>(type: K, listener: (this: HTMLInputElement, ev: HTMLElementEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
```

The main difference being that I'm adding in an extra generic type onto the callback.

**Comment by webstrand** [comment]
I am unable to reproduce your issue in either typescript 3.5.3 called from a terminal or in the [playground](https://www.typescriptlang.org/play/#code/PTAEEFQKgZwUwDYDMC0BjA9gOwC4EMBLLOAEylBLgFtsYcAnPHA7UDJUHACzlAAd6GAEYJqoJBgQIMAdxgA6RQCgQoACpw6nLgRighAT1D0ArlixEA5qAAGOGGhtss23pipU8WEigRE4ADTGTDz02l7agiaWXKBeGNxwYUImBAgknBiSoDAmaLF4egDiJgh8QQDqcEJ8eGgA1kFwOGjySnAAHnwY9DigRDhJSHW8AJIAwvQdALIYlAgAogBucLgAPGoAfKAA3kqgB6AA5PlelnBHAFygAGp4CCZw4-cIQnX1G5sA3PuHJ1xnOCjXBJLD3K63e6PZ5SN4NT4-Q7HU5Yc4LDqDehghAQu4PJ4vOEfLY-AC+SiUOAMfF4eOhhPen1AAF5QAAKJZQuDXNQAShZ2yWGAIJB+7S6PT6eCEdEYaD6aAQhT0kxmc0QTL2SLwJBIy1WODWAGlQJ1Bt49PU4AZ2KAJlNZvN9estps2VSadcjUE0AyGtd7WqnSsXZsANpGgC6-L2vwO5PJ4u6vVAiuVoFVADkTFQhElTRjViQVQ71Qg1lgc3n6NsdqBE6I+pWqCzQMQZBmptnc0k2byfs35Dq9SGcGz-oCjkElgKIABlOcLABKajWsqsbqWvP7FMoafovCQZnlLBc4AXy9XWw5PN512IK3oXyAA). The type of parameter `v` is definitely inferred as `number` (otherwise the `ASSERT` in my example wouldn't produce an error). 

Can you reproduce your issue in the playground and provide a link for us? Or can it only be reproduced within a specific environment?

**Comment by RyanCavanaugh** [comment]
There's no definition for `CrxValidationState` here and trying to simplify the repro doesn't display the problem described. Can you reduce this to something that can be demonstrated in the Playground? Thanks

**Comment by sam-s4s** [comment]
@webstrand thank you for putting it in the playground! I have just opened your playground, and as you can see - I have hovered my mouse over `v` (without changing your example at all) and it is inferred as `any`.

![image](https://user-images.githubusercontent.com/18692089/63474160-67bed980-c4bb-11e9-9c3f-00236732868d.png)

@RyanCavanaugh ah sorry, `CrxValidationState` wasn't supposed to be there, I'll delete that out of the original post.

**Comment by fatcerberus** [comment]
![image](https://user-images.githubusercontent.com/4097841/63474266-8767ea80-c446-11e9-911a-efa8fd3eb6ef.png)

**Comment by sam-s4s** [comment]
Sorry, I was reading the wrong part like a derp. I'll have a play and then get back to you :)

**Comment by RyanCavanaugh** [comment]
Sounds good, please unassign if you narrow it down to something, or close if it turned out to be not-a-bug 🙂

**Comment by sam-s4s** [comment]
Sure thing - of note, I've pasted the playground example into my environment and it's correctly assigning `number`... so there must be another factor in my real code - an extra complexity...

**Comment by sam-s4s** [comment]
Ok, so I've managed to get a cut-down version that has the problem on my machine...
https://www.typescriptlang.org/play/#code/PTAEEFQKgZwUwDYDMC0BjA9gOwC4EMBLLOAEylBLgFtsYcAnPHA7UDJUHACzlAAd6GAEYJqoJBgQIMAdxgA6RQCgQoACpw6nLgRighAT1D0ArlixEA5qAAGOGGhtss23pipU8WEigRE4ADTGTDz02l7agiaWXKBeGNxwYUImBAgknBiSoDAmaLF4egDiJgh8QQDqcEJ8eGgA1kFwOGjySkpEOElIdbwAkgDC9AAeALIYlAgAogBucLgAPGoAfKAA3kqgW6AA5PlelnA7AFygAGp4CCZwA5cIQnX1S8sA3Jvbe1wHcH24SViXE7nS7XW5SB4NZ5vba7fZYQ5TYZdegAhBAi5XG53CFPFbQj4zS4EEhMI6nDHEpgsLBg+6PF6gVTuKjzHDaXSgPzENgmNleDIENkyDD0ep6TD0ehwNA4BAGJQAX3aOAMfF4GNB2Mez1AAF5QAAKQmY05qACUetWMwwxLeKrVwL8JOY2FpOL1huNlJdWAAyvguqchFlRF4LbqrTaSG8lHghHRGDLQGgEIU9ENhgNsAxsmtQErY-GGHU2Sm06AM+NJjq4Ej5iR0yMs7hBAh1u8tngSCRZqyFgBpUC1rrePT1OAGdigQYjKuIXuLFbLA32uCnftBNBahqnGdjCbzuaL5YAbX7AF0LRsO-nFe0yzBG8MAHImKhCJJDuujiuzg8IBYsDfD96FWPMC1ENkgKoD1iBkX8X2ApIDTNN5oPkLseyPHADU+b4diCGZLXWfNUKUIA

However, in the Playground it works correctly! Ahhh! What's the difference? Same piece of code.

Playground is running 3.5.1 and I'm running 3.5.3, but surely that's not the problem.

When I comment out the `'validate'` line in the `ICrxModelEvent` the problem goes away, but when it's included, the problem is present. But... not in the playground :(

**Comment by sam-s4s** [comment]
Oh noes, how embarrassment...

So this problem first came up when I was running Angular 7, which uses Typescript 3.2.4. It presented both in Visual Studio 2017's intellisense, and in the Typescript compiler.

Yesterday I upgraded to Angular 8, which uses Typescript 3.5.3, and checked the problem. Intellisense still reported the problem, so off to github I went. Sadly I forgot that Visual Studio 2017 doesn't use the project's version of Typescript, but it's own weird internal version.

So, in short, this problem was fixed somewhere between 3.2.4 and 3.5.1, and is now only presenting in Visual Studio 2017. Sorry! Looks like I just need to figure out how to upgrade TS in VS 2017.

I shall close the case :)

Thank you for your help everyone!
